lambdas/handlers/authoriser_handler.py

```diff
@@ -0,0 +1,268 @@
+        ssm_response = client.get_parameter(
+            Name=ssm_public_key_parameter_name, WithDecryption=True
+        )
+        public_key = ssm_response["Parameter"]["Value"]
+
+        decoded = jwt.decode(
+            event["authorizationToken"], public_key, algorithms=["RS256"]
+        )
+
+        ndr_session_id = decoded["ndr_session_id"]
+
+        current_session = find_login_session(ndr_session_id)
+        validate_login_session(current_session, ndr_session_id)
+
+        # if user has a valid session, assign their role
+        user_roles = [org["role"] for org in decoded["organisations"]]
+
+    except AuthorisationException as e:
+        logger.error(e)
+    except botocore.exceptions.ClientError as e:
+        logger.error(e)
+    except jwt.PyJWTError as e:
+        logger.error(f"error while decoding JWT: {e}")
+    except (KeyError, IndexError) as e:
+        logger.error(e)
+
+    principal_id = ""
+    _, _, _, region, aws_account_id, api_gateway_arn = event["methodArn"].split(":")
+    api_id, stage, http_verb, resource_name = api_gateway_arn.split("/")
+
+    policy = AuthPolicy(principal_id, aws_account_id)
+    policy.restApiId = api_id
+    policy.region = region
+    policy.stage = stage
+
+    # for now, allow all method for GP and DEV role, and allow only search document for PCSE
+    if PermittedRole.DEV.name in user_roles:
+        policy.allowAllMethods()
+    elif PermittedRole.GP.name in user_roles:
+        policy.allowAllMethods()
+    elif PermittedRole.PCSE.name in user_roles:
+        policy.allowMethod(HttpVerb.GET, "/SearchDocumentReferences")
+    else:
+        policy.denyAllMethods()
+
+    auth_response = policy.build()
+
+    return auth_response
+
+
+def redact_id(session_id: str) -> str:
+    # Extract the last 4 chars of session id for logging, as it was in ARF
+    return session_id[-4:]
+
+
+def find_login_session(ndr_session_id):
+    logger.debug(
+        f"Retrieving session for session ID ending in: f{redact_id(ndr_session_id)}"
+    )
+    session_table_name = os.environ["AUTH_SESSION_TABLE_NAME"]
+    db_service = DynamoDBService()
+    query_response = db_service.simple_query(
+        table_name=session_table_name,
+        key_condition_expression=Key("NDRSessionId").eq(ndr_session_id)
+    )
+
+    try:
+        current_session = query_response["Items"][0]
+        return current_session
+    except (KeyError, IndexError) as error:
+        logger.info(error)
+        raise AuthorisationException(
+            f"Unable to find session for session ID ending in: {redact_id(ndr_session_id)}"
+        )
+
+
+def validate_login_session(current_session, ndr_session_id):
+    expiry_time = current_session["TimeToExist"]
+    time_now = time.time()
+    if expiry_time <= time_now:
+        raise AuthorisationException(
+            f"The session is already expired for session ID ending in: {redact_id(ndr_session_id)}"
+        )
+
+
+class HttpVerb:
```

Comment:
This file feels like it's getting pretty big with lots of components for just a lambda handler, maybe could extract these out into utils or models to clean up the file slightly?

Reply:
We kept those class in the lambda handler because those are from AWS blueprint so we tried to keep them in one place, but I am happy to move the classes to a separate file.